### ISSUE TITLE
Incorrect interpretation of RFC 6265 in SetCookie::validate

### ISSUE THREAD

**Issue by Naktibalda** [guzzle/guzzle#1205]
[SetCookie::validate](https://github.com/guzzle/guzzle/blob/6692db3cda013e8357330243c33fe1838d0ea6f5/src/Cookie/SetCookie.php#L366) checks if the cookie is valid according to RFC 6265

It includes the following checks:
  1. Names must not be empty, but can be 0
  1. Value must not be empty, but can be 0

[RFC 6265](http://tools.ietf.org/html/rfc6265#page-18)
```
   3.  The (possibly empty) name string consists of the characters up
       to, but not including, the first %x3D ("=") character, and the
       (possibly empty) value string consists of the characters after
       the first %x3D ("=") character.

   4.  Remove any leading or trailing WSP characters from the name
       string and the value string.

   5.  If the name string is empty, ignore the set-cookie-string
       entirely.

```

It only says to ignore the cookie, not to do something drastic like throwing exception, when name string is empty and says nothing about empty value.

**Comment by mtdowling** [comment]
Sounds good to me. Can you send a PR?

**Comment by Naktibalda** [comment]
I solved my issue by disabling strict mode.

I don't know what is the intention of strict mode in CookieJar.
Changing comment would count as a fix, right?
```diff
- * Check if the cookie is valid according to RFC 6265
+ * Check if the cookie is valid
```

**Comment by mtdowling** [comment]
I don't think that is the best way to fix it. The validate check should not be called when there is no name.

**Comment by Naktibalda** [comment]
A related issue is caused by this condition:
```php
        // Check if any of the invalid characters are present in the cookie name
        if (preg_match('/[\x00-\x20\x22\x28-\x29\x2c\x2f\x3a-\x40\x5b-\x5d\x7b\x7d\x7f]/', $name)) {
            return 'Cookie name must not contain invalid characters: ASCII Control characters (0-31;127), space, tab and the following characters: ()<>@,;:\"/[]?={}';
        }
```

I have no objections against rejecting most of these characters, but 2 of them are very important - square brackets are used to define arrays, even PHP manual gives example how to do that http://php.net/setcookie#example-4936

**Comment by mtdowling** [comment]
Addressed by #1245

**Comment by pakulin** [comment]
Run into non standard cookie-name with ':' char, and need to override SetCookie::class but there's a bit tight coupling in CookieJarInterface
